espnet/nets/pytorch_backend/k2_ctc.py

```diff
@@ -0,0 +1,150 @@
+from typing import Tuple
+import logging
+
+
+class K2CTCLoss(torch.nn.Module):
+
+    def __init__(self, odim: int, reduction: str = 'sum') -> None:
+        torch.nn.Module.__init__(self)
+        self.graph_compiler = CtcTrainingGraphCompiler(odim)
+
+    def forward(self, log_probs: torch.Tensor, targets: torch.Tensor,
+                input_lengths: torch.Tensor,
+                target_lengths: torch.Tensor) -> torch.Tensor:
+
+        log_probs = log_probs.permute(
+            1, 0, 2
+        )  # now log_probs is [N, T, C]  batchSize x seqLength x alphabet_size
+        supervision_segments = torch.stack(
+            (torch.tensor(range(input_lengths.shape[0])),
+             torch.zeros(input_lengths.shape[0]), input_lengths),
+            1).to(torch.int32)
+        #indices = torch.argsort(supervision_segments[:, 2], descending=True)
+        #supervision_segments = supervision_segments[indices]
+
+        dense_fsa_vec = k2.DenseFsaVec(log_probs, supervision_segments)
+
+        decoding_graph = self.graph_compiler.compile(
+            targets, target_lengths).to(log_probs.device)
+        # decoding_graph = k2.index(decoding_graph,
+        #                          indices.to(torch.int32)).to(log_probs.device)
+        # assert decoding_graph.device == log_probs.device
+
+        target_graph = k2.intersect_dense(decoding_graph, dense_fsa_vec, 10.0)
+        tot_scores = target_graph.get_tot_scores(log_semiring=True,
+                                                 use_double_scores=False)
+        # tot_scores = tot_scores / target_lengths.to(log_probs.device)
+        (tot_score, tot_frames,
+         all_frames) = get_tot_objf_and_num_frames(tot_scores,
+                                                   supervision_segments[:, 2])
+        return -tot_score
+
+
+def build_ctc_topo(tokens: List[int]) -> k2.Fsa:
+    '''Build CTC topology.
+    A token which appears once on the right side (i.e. olabels) may
+    appear multiple times on the left side (ilabels), possibly with
+    epsilons in between.
+    When 0 appears on the left side, it represents the blank symbol;
+    when it appears on the right side, it indicates an epsilon. That
+    is, 0 has two meanings here.
+    Args:
+      tokens:
+        A list of tokens, e.g., phones, characters, etc.
+    Returns:
+      Returns an FST that converts repeated tokens to a single token.
+    '''
+    assert 0 in tokens, 'We assume 0 is ID of the blank symbol'
+
+    num_states = len(tokens)
+    final_state = num_states
+    arcs = ''
+    for i in range(num_states):
+        for j in range(num_states):
+            if i == j:
+                arcs += f'{i} {i} {tokens[i]} 0 0.0\n'
+            else:
+                arcs += f'{i} {j} {tokens[j]} {tokens[j]} 0.0\n'
+        arcs += f'{i} {final_state} -1 -1 0.0\n'
+    arcs += f'{final_state}'
+    ans = k2.Fsa.from_str(arcs)
+    ans = k2.arc_sort(ans)
+    return ans
+
+
+class CtcTrainingGraphCompiler(object):
```

Comment:
We have another version of building graphs where all operations are moved to CUDA.
Please see 
https://github.com/k2-fsa/snowfall/blob/2e679045be6fc251e9fbb5fff614a5e35104136f/snowfall/training/mmi_mbr_graph.py#L109

Reply:
There may be problems with multi-gpu training if we build graphs with CUDA and put graphs in a specific GPU memories. I haven't figured out how to deal with this problem, but I don't think it's a problem because of the speed of building graphs is quite faster and the size of CTC graph is small enough.

Reply:
mm I didn't like setting the gpu_id explicitly either.  I feel like we should be getting it from somewhere else.

Reply:
Anyway, I have implemented the CUDA graphs building method. But it will not be used until I figure out how to work with multi-gpu training.